### PR TITLE
MAKE-745: cli interface

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: cli/reporting.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+func reportingJobStatus(opts *ServiceOptions) cli.Command {
+	return cli.Command{
+		Name: "status",
+		Flags: opts.reportingFlags(
+			cli.StringFlag{
+				Name:  "filter",
+				Value: "in-progress",
+				Usage: "specify the process filter, can be 'in-progress' (default), 'pending', or 'stale'",
```
Comment (Kimchelly) [inline cli/reporting.go]: The "default" is superfluous since the cli should already print the default when Value is set.

File: cli/management.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	}
+}
+
+func manageCheckJob(opts *ServiceOptions) cli.Command {
+	const jobIDFlagName = "id"
+
+	return cli.Command{
+		Name: "check",
+		Flags: opts.managementFlags(
+			cli.StringSliceFlag{
```
Comment (Kimchelly) [inline cli/management.go]: Should there be a check to ensure this flag is set?

File: cli/management.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	}
+}
+
+func manageAbortJob(opts *ServiceOptions) cli.Command {
+	const jobIDFlagName = "id"
+
+	return cli.Command{
+		Name: "abort",
+		Flags: opts.managementFlags(
+			cli.StringSliceFlag{
```
Comment (Kimchelly) [inline cli/management.go]: Same question as above.

File: cli/reporting.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+		Flags: opts.reportingFlags(
+			cli.DurationFlag{
+				Name:  "duration, d",
+				Value: time.Minute,
+				Usage: "specify a duration in string form (e.g. 100ms, 1s, 1m, 1h) to limit the report",
+			},
+			cli.StringFlag{
+				Name:  "filter",
+				Value: "completed",
+				Usage: "specify the runtime filter, can be 'completed' (default), 'latency', or 'running'",
```
Comment (Kimchelly) [inline cli/reporting.go]: Same as above - no need to explicitly mention default.

File: cli/reporting.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+func reportingJobIDs(opts *ServiceOptions) cli.Command {
+	return cli.Command{
+		Name: "jobs",
+		Flags: opts.reportingFlags(
+			cli.StringFlag{
+				Name:  "filter",
+				Value: "in-progress",
+				Usage: "specify the process filter, can be 'in-progress' (default), 'pending', or 'stale'",
```
Comment (Kimchelly) [inline cli/reporting.go]: See above about default.

File: cli/reporting.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		Flags: opts.reportingFlags(
+			cli.DurationFlag{
+				Name:  "duration, d",
+				Value: time.Minute,
+				Usage: "specify a duration in string form (e.g. 100ms, 1s, 1m, 1h) to limit the report",
+			},
+			cli.StringFlag{
+				Name:  "filter",
+				Value: "unique-errors",
+				Usage: "specify the process filter, can be 'unique-errors' (default), 'all-errors', or 'stats-only'",
```
Comment (Kimchelly) [inline cli/reporting.go]: See default comment.

File: cli/reporting.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+						if err != nil {
+							return errors.WithStack(err)
+						}
+						reports = append(reports, report)
+					}
+				}
+
+				t := tabby.New()
+				t.AddHeader("Job Type", "Count", "Total", "Average", "First Error", "Group")
+				for _, report := range reports {
```
Comment (bsamek) [inline cli/reporting.go]: slice of pointers

File: cli/cli.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	Client           *http.Client
+}
+
+const (
+	serviceURLFlagName = "service"
+	prefixFlagName     = "prefix"
+)
+
+func (o *ServiceOptions) reportingFlags(base ...cli.Flag) []cli.Flag {
+	return append(base, []cli.Flag{
```
[the remark this answers is not in the document]
Comment (tychoish) [reply cli/cli.go]: I think you have to unpack base or the new slice, either way.

File: cli/cli.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+		cli.StringFlag{
+			Name:  prefixFlagName,
+			Usage: "specify the service prefix for the reporting service",
+			Value: o.ReportingPrefix,
+		},
+	}...)
+}
+
+func (o *ServiceOptions) managementFlags(base ...cli.Flag) []cli.Flag {
+	return append(base, []cli.Flag{
```
[the remark this answers is not in the document]
Comment (tychoish) [reply cli/cli.go]: something has to be unpacked.

File: cli/cli.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		},
+	}...)
+}
+
+func (o *ServiceOptions) withReportingClient(ctx context.Context, c *cli.Context, op func(client *rest.ReportingClient) error) error {
+	if o.Client == nil {
+		o.Client = http.DefaultClient
+	}
+
+	url := c.String(serviceURLFlagName) + c.String(prefixFlagName)
```
[the remark this answers is not in the document]
Comment (tychoish) [reply cli/cli.go]: done!

File: cli/management.go (hunk 2 above)
Comment (tychoish) [reply cli/management.go]: It'll 404, and there's only so much client side validation we can really provide.

File: cli/reporting.go (hunk 1 above)
Comment (tychoish) [reply cli/reporting.go]: done.

File: cli/reporting.go (hunk 7 above)
Comment (tychoish) [reply cli/reporting.go]: addline calls fprintf on its inputs, so I think we're ok here

Comment (bsamek) [reply cli/reporting.go]: I suppose we'll find out quickly if there's a problem.

**Pull request merged**